client/asset/eth/contractor.go

```diff
@@ -194,7 +213,17 @@ func (c *contractorV0) swap(ctx context.Context, secretHash [32]byte) (*dexeth.S
 
 // refund issues the refund command to the swap contract. Use isRefundable first
 // to ensure the refund will be accepted.
-func (c *contractorV0) refund(txOpts *bind.TransactOpts, secretHash [32]byte) (*types.Transaction, error) {
+func (c *contractorV0) refund(txOpts *bind.TransactOpts, secretHash [32]byte) (tx *types.Transaction, err error) {
+
+	// If refund fails with a provider, trust the next nonce.
+	if mRPC, is := c.cb.(*multiRPCClient); is {
+		defer func() {
+			if err != nil {
+				mRPC.voidUnusedNonce()
+			}
+		}()
+	}
```

Comment:
The defer would fire to early.

Reply:
I believe it should work with [that](https://github.com/golang/go/issues/49203) ?

Here is an experiment that seems to be working same (before and after suggested deduplication): https://go.dev/play/p/aswoXUntuLD

Reply:
I know when things are evaluated with defers.  I thought you were suggesting to just put the entire `if` block in a helper function.
But please don't make that de-dup change.  That kind of contortion to avoid a few lines of code is a net negative.  You may propose an actual change @norwnd with this specific code, not toy code, and we can discuss that.

Reply:
Could just remove the inline error checks I guess.  mRPC.voidUnusedNonce() could take an error argument and do nothing if nil. I don't know if that makes the code clearer or more convoluted. A separate helper function would be clear enough but not save so many lines. Was waiting for consensus if the node cache was even a good idea. can optimize later, imo